cpp/order/RotationalAutocorrelation.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2010-2018 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include "RotationalAutocorrelation.h"
+
+#include <math.h>
+
+/*! \file RotationalAutocorrelation.cc
+    \brief Implements the RotationalAutocorrelation class.
+*/
+
+namespace freud { namespace order {
+
+// This convenience function wraps exponentiation for complex numbers to avoid
+// the case where pow((0,0), 0) returns (nan, nan).
+std::complex<float> cpow(std::complex<float> base, float p)
+    {
+    if (p == 0)
+        {
+        return std::complex<float> (1, 0);
+        }
+    else
+        {
+        return std::pow(base, p);
+        }
+    }
+
+// This convenience function wraps std's gamma function for factorials, with
+// the appropriate shift by 1.
+float factorial(int n)
+    {
+    return std::tgamma(n+1);
+    }
+
+std::pair<std::complex<float>, std::complex<float> > quat_to_greek(const quat<float> &q)
+    {
+    std::complex<float> xi(q.v.x, q.v.y);
+    std::complex<float> zeta(q.v.z, q.s);
+    return std::pair<std::complex<float>, std::complex<float> >(xi, zeta);
+    }
+
+
+std::complex<float> hypersphere_harmonic(const std::complex<float> xi, std::complex<float> zeta,
+                                         const int l, const int m1, const int m2)
+    {
+    const int a = -(m1 - l/2);
+    const int b = -(m2 - l/2);
+
+    // Doing a summation over non-negative exponents, which requires the additional inner conditional.
+    std::complex<float> sum_tracker(0,0);
+    for (int k = 0; k <= std::min(a, b); k++)
+        {
+        if (l + k - a - b >= 0)
+            {
+            sum_tracker += cpow(std::conj(xi),k) * cpow(zeta, b-k) *
+                              cpow(std::conj(zeta), a-k) * cpow(-xi, l+k-a-b) /
+                              factorial(k) / factorial(l+k-a-b) /
+                              factorial(a-k) / factorial(b-k);
+            }
+        }
+    sum_tracker *= std::sqrt(factorial(a) * factorial(l-a) * factorial(b) * factorial(l-b) / (float(l)+1));
```

Comment:
I'm wondering if there is a more numerically stable way to do this... Lots of (large?) factorial computations with `float` precision.

Reply:
That's a fair question. The factorial argument is bounded by l, which I assume will typically be in the single digits. I'm not sure there's a good way around this without trying to rework the underlying math, can you think of anything?

Reply:
I don't think there is a clear way to mathematically simplify this expression. I considered proposing `lgamma` but it's not thread-safe and probably doesn't matter much if `l` is a small integer. We could consider switching to `double` precision for this summation, or leaving it as-is. Just wanted to think about it a bit.
https://en.cppreference.com/w/cpp/numeric/math/lgamma